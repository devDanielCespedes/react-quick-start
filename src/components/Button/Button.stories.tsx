import type { Meta, StoryObj } from "@storybook/react";
import { Button } from "./Button";
import { fn } from "@storybook/test";
import { withActions } from "@storybook/addon-actions/decorator";

const meta: Meta<typeof Button> = {
  title: "Components/Button",
  component: Button,
  tags: ["autodocs"],
  parameters: {
    layout: "centered",
    actions: { argTypesRegex: "^on.*", handles: ["mouseover", "click .btn"] },
  },
  argTypes: {
    color: {
      control: "select",
      options: ["primary", "secondary", "success", "error", "warning", "info"],
    },
    size: { control: "select", options: ["small", "medium", "large"] },
    variant: { control: "select", options: ["text", "outlined", "contained"] },
    loading: { control: "boolean" },
    disabled: { control: "boolean" },
    onClick: fn(),
  },
  decorators: [withActions],
};

export default meta;
type Story = StoryObj<typeof meta>;

export const Contained: Story = {
  args: {
    children: "Contained Button",
    variant: "contained",
    color: "primary",
  },
};

export const Text: Story = {
  args: {
    children: "Text Button",
    variant: "text",
    color: "primary",
  },
};

export const Loading: Story = {
  args: {
    children: "Loading Button",
    loading: true,
    variant: "contained",
    color: "primary",
  },
};

export const Disabled: Story = {
  args: {
    children: "Disabled Button",
    variant: "contained",
    color: "primary",
    disabled: true,
  },
};

export const Small: Story = {
  args: {
    children: "Small Button",
    size: "small",
    variant: "contained",
  },
};

export const Medium: Story = {
  args: {
    children: "Medium Button",
    size: "medium",
    variant: "contained",
  },
};

export const Large: Story = {
  args: {
    children: "Large Button",
    size: "large",
    variant: "contained",
  },
};

export const Outlined: Story = {
  args: {
    children: "Outlined Button",
    variant: "outlined",
    color: "primary",
  },
};

export const PrimaryColor: Story = {
  args: {
    children: "Primary Color Button",
    variant: "contained",
    color: "primary",
  },
};

export const SecondaryColor: Story = {
  args: {
    children: "Secondary Color Button",
    variant: "contained",
    color: "secondary",
  },
};

export const SuccessColor: Story = {
  args: {
    children: "Success Color Button",
    variant: "contained",
    color: "success",
  },
};

export const ErrorColor: Story = {
  args: {
    children: "Error Color Button",
    variant: "contained",
    color: "error",
  },
};

export const WarningColor: Story = {
  args: {
    children: "Warning Color Button",
    variant: "contained",
    color: "warning",
  },
};

export const InfoColor: Story = {
  args: {
    children: "Info Color Button",
    variant: "contained",
    color: "info",
  },
};

export const CustomColor: Story = {
  args: {
    children: "Custom Color Button",
    variant: "contained",
    sx: { backgroundColor: "#EC9021", color: "#fff" },
  },
};

export const WithIcon: Story = {
  args: {
    children: "With Icon",
    variant: "contained",
    startIcon: (
      <span role="img" aria-label="icon">
        ðŸš€
      </span>
    ),
  },
};

export const IconOnly: Story = {
  args: {
    variant: "contained",
    startIcon: (
      <span role="img" aria-label="icon">
        ðŸš€
      </span>
    ),
  },
};
