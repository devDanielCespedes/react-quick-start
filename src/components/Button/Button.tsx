import React from "react";
import LoadingButton, { LoadingButtonProps } from "@mui/lab/LoadingButton";
import { styles } from "./styles";

const Button = React.forwardRef<HTMLButtonElement, LoadingButtonProps>(
  (props, ref) => {
    return <LoadingButton sx={styles} {...props} ref={ref} />;
  }
);

Button.displayName = "Button";

export { Button };
