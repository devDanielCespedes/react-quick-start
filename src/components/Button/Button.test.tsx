import { render, screen, fireEvent } from "@testing-library/react";
import { Button } from "./Button";

test("renders Button with text", () => {
  const { asFragment } = render(<Button>Test Button</Button>);
  const buttonElement = screen.getByText(/Test Button/i);
  expect(buttonElement).toBeInTheDocument();
  expect(asFragment()).toMatchSnapshot();
});

test("Button onClick works", () => {
  const handleClick = jest.fn();
  const { asFragment } = render(
    <Button onClick={handleClick}>Click Me</Button>
  );
  const buttonElement = screen.getByText(/Click Me/i);
  fireEvent.click(buttonElement);
  expect(handleClick).toHaveBeenCalledTimes(1);
  expect(asFragment()).toMatchSnapshot();
});
