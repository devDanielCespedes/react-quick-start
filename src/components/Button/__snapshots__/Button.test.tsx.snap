// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button onClick works 1`] = `
<DocumentFragment>
  <button
    class="MuiButtonBase-root MuiButton-root MuiLoadingButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiLoadingButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-18wz6ak-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
    id=":r1:"
    tabindex="0"
    type="button"
  >
    Click Me
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</DocumentFragment>
`;

exports[`renders Button with text 1`] = `
<DocumentFragment>
  <button
    class="MuiButtonBase-root MuiButton-root MuiLoadingButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiLoadingButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-18wz6ak-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
    id=":r0:"
    tabindex="0"
    type="button"
  >
    Test Button
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</DocumentFragment>
`;
